snakemake/logging.py

```
@@ -305,12 +243,504 @@
+        fmt_precision = "{{:.{}%}}".format
+        fmt = lambda fraction: fmt_precision(precision).format(fraction)
+        while fmt(fraction) == "100%" or fmt(fraction) == "0%":
+            precision += 1
+        return fmt(fraction)
+
+
+def show_logs(logs):
+    """Helper method to show logs."""
+    for f in logs:
+        try:
+            content = open(f, "r").read()
```

Ruff

Use a context manager for opening files

(SIM115)